src/List.mo

```diff
@@ -1802,14 +2101,43 @@ module {
   ///
   /// Space: `O(1)`
   public func reverse<T>(list : List<T>) : List<T> {
-    let rlist = empty<T>();
+    let rlist = repeatInternal<T>(null, size(list));
 
-    reverseForEach<T>(
-      list,
-      func(x) = add(rlist, x)
-    );
+    let blocks = list.blocks.size();
+    var blockIndexFront = 0;
+    var elementIndexFront = 0;
+    var sz = 0;
+    var dbFront : [var ?T] = [var];
+
+    var blockIndexBack = rlist.blockIndex;
+    var elementIndexBack = rlist.elementIndex;
+    var dbBack : [var ?T] = if (blockIndexBack < rlist.blocks.size()) {
+      rlist.blocks[blockIndexBack]
+    } else { [var] };
+
+    loop {
+      if (elementIndexFront == sz) {
+        blockIndexFront += 1;
+        if (blockIndexFront >= blocks) return rlist;
+        dbFront := list.blocks[blockIndexFront];
+        sz := dbFront.size();
+        if (sz == 0) return rlist;
+        elementIndexFront := 0
+      };
 
-    rlist
+      if (elementIndexBack == 0) {
+        blockIndexBack -= 1;
+        if (blockIndexBack == 0) return rlist;
+        dbBack := rlist.blocks[blockIndexBack];
+        elementIndexBack := dbBack.size() - 1
+      } else {
+        elementIndexBack -= 1
+      };
+
+      dbBack[elementIndexBack] := dbFront[elementIndexFront];
+
+      elementIndexFront += 1
+    }
```

Comment:
Since we are looping over the entire list (not half the list as in reverseInPlace) can't we use the normal, better known forward-moving while loop that we use in all other places across the package? That would be more familiar code. Then we only need to move the two `..Back` indices backwards inside the while loop.